src/brevitas/graph/equalize.py

```diff
@@ -196,22 +293,44 @@ def walk_region(graph_model: GraphModule, starting_node: Node, history, srcs, si
             continue
         if _is_supported_module(graph_model, node):
             if walk_forward:
-                sinks.add(node.target)
+                if node.target not in sinks:
+                    sinks[node.target] = 0
             else:
-                srcs.add(node.target)
-                walk_region(graph_model, node, history, srcs, sinks, walk_forward=True)
+                if node.target not in srcs:
+                    srcs[node.target] = len(srcs.keys())
+                walk_region(graph_model, node, history, srcs, sinks, walk_forward=True, seen_concat=seen_concat)
         elif _is_scale_invariant_module(graph_model, node):
             if walk_forward:
-                walk_region(graph_model, node, history, srcs, sinks, walk_forward=True)
+                walk_region(graph_model, node, history, srcs, sinks, walk_forward=True, seen_concat=seen_concat)
             else:
-                walk_region(graph_model, node, history, srcs, sinks, walk_forward=True)
-                walk_region(graph_model, node, history, srcs, sinks, walk_forward=False)
+                walk_region(graph_model, node, history, srcs, sinks, walk_forward=True, seen_concat=seen_concat)
+                walk_region(graph_model, node, history, srcs, sinks, walk_forward=False, seen_concat=seen_concat)
         elif (node.op == 'call_method' and node.target in _residual_methods
             or node.op == 'call_function' and node.target in _residual_fns):
-                walk_region(graph_model, node, history, srcs, sinks, walk_forward=True)
-                walk_region(graph_model, node, history, srcs, sinks, walk_forward=False)
+                walk_region(graph_model, node, history, srcs, sinks, walk_forward=True, seen_concat=seen_concat)
+                walk_region(graph_model, node, history, srcs, sinks, walk_forward=False, seen_concat=seen_concat)
+        elif node.target in _cat_fns:
+            if seen_concat is not None:
+                # If we go through another concat, then no equalization should be done.
+                # Removing all the sinks should guarantee that.
+                _clear_sinks(sinks)
+                continue
+            # If this is the first concat we see, we register it while we keep exploring the graph.
+            # NOTE: seen_concat=node is better than seen_concat=True if we decide to support
+            # partial equalization through concats. No difference otherwise.
+
+            # If any of the input to cat has multiple users, we cannot equalize through
+            for n in node.all_input_nodes:
+                if len(n.users)>1:
+                    return
```

Comment:
Not sure if here we should do `continue` of the outer loop or return or it's the same thing @volcacius